### ISSUE TITLE
Debug isn't working (optimized out)

### ISSUE THREAD

**Issue by JCKodel** [Dart-Code/Dart-Code#4555]
**Describe the bug**
During debug, hovering a value doesn't show its value anymore.

The code is running on debug on an Android device.

**To Reproduce**
1. Run some code
2. Break in a breakpoint then hover some variable

**Expected behavior**
To show the value of the variable.

**Screenshots**
![image](https://github.com/Dart-Code/Dart-Code/assets/379339/3334f4be-66b6-4342-9600-3b931d32c1d2)

**Please complete the following information:**
 - Operating System and version: Windows 11 22H2
 - VS Code version: 1.78.2
 - Dart extension version: 3.64.0
 - Dart/Flutter SDK version: Daert 3.0.1, Flutter 3.10.1
 - Target device (if the issue relates to Flutter debugging): Android emulator

**Comment by DanTup** [comment]
@noga-dev are you able to give a specific code example so I can be sure I'm trying with something that triggers it for you? It's not clear to me if I couldn't reproduce it above because there's something different about the code I used, or if something else is a factor.

The code I used was:

```
final a = Completer();
print(a.isCompleted);
```

I inserted it into the `build` method of a new project created with **Flutter: New Project** and I hovered over both `a` and `isCompleted` in the `print` statement.

Can you also confirm your OS and exact Dart/Flutter versions? Thanks!

**Comment by noga-dev** [comment]
> @noga-dev are you able to give a specific code example so I can be sure I'm trying with something that triggers it for you? It's not clear to me if I couldn't reproduce it above because there's something different about the code I used, or if something else is a factor.
> 
> The code I used was:
> 
> ```
> final a = Completer();
> print(a.isCompleted);
> ```
> 
> I inserted it into the `build` method of a new project created with **Flutter: New Project** and I hovered over both `a` and `isCompleted` in the `print` statement.
> 
> Can you also confirm your OS and exact Dart/Flutter versions? Thanks!

<img width="1280" alt="Code_TSG0ylxCax" src="https://github.com/Dart-Code/Dart-Code/assets/2067558/9dfa0bdd-9606-400f-ae40-c3a0a61116b8">

![image](https://github.com/Dart-Code/Dart-Code/assets/2067558/7710e437-ca1c-411d-b4cc-bcbb4f360fb2)

Looks good with a new project.

Win11. Flutter 3.10.1. Dart 3.0.1.

**Comment by DanTup** [comment]
> Guess it must have been a VSC glitch?

I dunno - based on the errors ("optmized our" and "invalid params") I feel like something is wrong in the Dart side (either the debug adapter, or the VM), but it's difficult to debug without a repro/log.

If you see it again, please let me know and provide steps. If you can reproduce it after restarting the debug session, then a log would be helpful to (although you need to start logging before starting the debug session):

- Run the **Dart: Capture Debugging Logs** command
- Start the debug session and reproduce the issue
- Click **Cancel** on the logging notification

Note: The log may contain data from your app so ensure there's nothing sensitive before sharing. It's likely only the line that contains "Optimized out" and a few lines before it are all that useful to me.

You can send logs to logs@dartcode.org (but still, don't send anything sensitive).

**Comment by ozalisky** [comment]
@DanTup I have the same issue
Here is the simple code to sign in with apple:
```dart
import 'package:firebase_auth/firebase_auth.dart';

                    final appleProvider = AppleAuthProvider()
                      ..addScope('email')
                      ..addScope('name');

                    await FirebaseAuth.instance
                        .signInWithProvider(appleProvider);
                    final user = FirebaseAuth.instance.currentUser;
                    print(user!.displayName);
                    print(user.email);
                    print(appleProvider);
```

![image](https://github.com/Dart-Code/Dart-Code/assets/40992789/26f2eccd-e455-44eb-828a-3485be6f810a)

VS Code
Version: 1.78.2
Commit: b3e4e68a0bc097f0ae7907b217c1119af9e03435
Date: 2023-05-10T15:00:40.428Z
Electron: 22.5.2
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Darwin arm64 22.4.0
Sandboxed: Yes

❯ flutter doctor -v
[✓] Flutter (Channel stable, 3.10.1, on macOS 13.3.1 22E772610a darwin-arm64, locale uk-UA)
    • Flutter version 3.10.1 on channel stable at /Users/oleksandrzaliskyi/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d3d8effc68 (7 days ago), 2023-05-16 17:59:05 -0700
    • Engine revision b4fb11214d
    • Dart version 3.0.1
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/oleksandrzaliskyi/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] VS Code (version 1.78.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.64.0

[✓] Connected device (3 available)
    • Oleksandr’s iPhone (mobile) • 00008110-0006499134B9801E                • ios          • iOS 16.5 20F66
    • John’s iPhone (mobile)      • ed8ce3983570a6e0b115365d11866d2b99868735 • ios          • iOS 15.7.5 19H332
    • macOS (desktop)             • macos                                    • darwin-arm64 • macOS 13.3.1 22E772610a darwin-arm64
      Xcode. Both devices may also need to be restarted and unlocked. (code 1)

[✓] Network resources
    • All expected network resources are available.

**Comment by DanTup** [comment]
@ozalisky can you reproduce this in any code that doesn't rely on Firebase/Apple Auth? I tried setting this up but I'm having issues making it work.

**Comment by DanTup** [comment]
Someone pinged me instructions and I was able to repro this inside a `FutureBuilder`'s callback:

<img width="674" alt="Screenshot 2023-05-25 at 12 09 16" src="https://github.com/Dart-Code/Dart-Code/assets/1078012/6f3eb76e-7c02-4fba-b599-0cd290da3074">

In the new SDK DAP it seems like we fail the `hover` request entirely, so the hover just doesn't show up instead. Both are essentially the same problem though. I'll try trim it down to something smaller.. I suspect it might be a VM/runtime issue.

**Comment by DanTup** [comment]
Smaller repro... I can only trigger it inside the `build` method (the one in `main` works):

```
import 'dart:developer';

import 'package:flutter/material.dart';

void main() {
  (int aaa) {
    debugger();
    print(aaa); // hover over aaa WORKS
  }(1);

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    (int aaa) {
      debugger();
      print(aaa); // hover over aaa DOES NOT WORK
    }(1);
    return const MaterialApp();
  }
}
```

Working VM request/response:

```
==> {
    "id": "695",
    "jsonrpc": "2.0",
    "method": "evaluateInFrame",
    "params": {
        "disableBreakpoints": true,
        "expression": "aaa",
        "frameIndex": 0,
        "isolateId": "isolates/708117027186575"
    }
}

<== {
    "jsonrpc": "2.0",
    "result": {
        "type": "@Instance",
        "_vmType": "Smi",
        "class": {
           // snip
        },
        "identityHashCode": 0,
        "kind": "Int",
        "fixedId": true,
        "id": "objects/int-1",
        "valueAsString": "1"
    },
    "id": "695"
}
```

Not working VM request/response:

```
==> {
    "id": "699",
    "jsonrpc": "2.0",
    "method": "evaluateInFrame",
    "params": {
        "disableBreakpoints": true,
        "expression": "aaa",
        "frameIndex": 0,
        "isolateId": "isolates/708117027186575"
    }
}

<== {
    "jsonrpc": "2.0",
    "result": {
        "type": "Sentinel",
        "kind": "OptimizedOut",
        "valueAsString": "<optimized out>"
    },
    "id": "699"
}
```

**Comment by DanTup** [comment]
There's an open issue in the SDK about this at https://github.com/dart-lang/sdk/issues/52430. I've added some additional info from above (I believe the issue will need to be fixed there).

**Comment by ndasim** [comment]
Same here, I'm even not able to watch local variables in the Android Studio when stopped in the breakpoint. Downgrading to 3.7.1 solves problem :(

**Comment by DanTup** [comment]
Please add 👍 to https://github.com/dart-lang/sdk/issues/52430 if you're affected by this. This will require a fix in the VM.

**Comment by DanTup** [comment]
This was fixed in the Dart SDK (via https://github.com/dart-lang/sdk/issues/52430) and the fix will be included in a future SDK release.

**Comment by Lonyless** [comment]
persists in Dart 3.1.0

**Comment by Lonyless** [comment]
```dart
CarouselSlider(
 items: items,
 carouselController: _slidersController.carouselController,
 options: CarouselOptions(
  viewportFraction: 1,
  onPageChanged: (index, __) {
    /// Happening inside onPageChanged
  },
  aspectRatio: 2.5,
  autoPlay: (items.length == 1 || items.isEmpty) ? false : true,
  autoPlayInterval: const Duration(seconds: 8),
  ),
);
``` 

Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.13.1, on Microsoft Windows [versÆo 10.0.19045.3324], locale pt-BR)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.6.3)
[√] Android Studio (version 2022.2)
[√] VS Code, 64-bit edition (version 1.81.1)
[√] Connected device (4 available)
[√] Network resources

![image](https://github.com/Dart-Code/Dart-Code/assets/48859696/74bc2d20-251f-4763-a191-eccb70ca509e)

**Comment by Lonyless** [comment]
did hide my comment because the problem seems unrelated to carousel_slider package.
it seems to be happening when i place a breakpoint in the build method of an widget that is inside a builder, wich is also inside a builder (two or more builders cascade)

![image](https://github.com/Dart-Code/Dart-Code/assets/48859696/104e98dd-4c49-4192-a17f-6b5ee0dba79b)

![image](https://github.com/Dart-Code/Dart-Code/assets/48859696/c6b183b9-5e20-4440-afde-cfe38a135d67)

**Comment by 2002Bishwajeet** [comment]
I still have this issue in some if else statements as well
<details>
<summary> Flutter Doctor </summary>

[✓] Flutter (Channel stable, 3.13.0, on macOS 13.4.1 22F770820d darwin-arm64 (Rosetta), locale en-IN)
    • Flutter version 3.13.0 on channel stable at /Users/biswa/.puro/envs/debug_env/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (9 days ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/biswa/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.5)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 75.1.3
    • Dart plugin version 231.9392.3

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.70.0
</details>

**Comment by DanTup** [comment]
@Lonyless I don't seem to be able to repro this, although it's also not clear to me what you're hovering over. Are you able to provide a complete code sample I can run, and an exact location/variable you are hovering over? Thanks!

**Comment by Lonyless** [comment]
@DanTup it's gone after 2 days. can't see what triggered it. At the time i tried dart pub cache repair, flutter clean and nothing worked. also tried doing a clean VSC install and nothing, guess it was some kind of cache problem?

**Comment by Lonyless** [comment]
@DanTup i noticed that the evaluation request registered in VSC did not match the expression i was trying to evaluate at the time

> banner.link!.allMatches('embed')

**Comment by DanTup** [comment]
> @DanTup it's gone after 2 days. can't see what triggered it. At the time i tried dart pub cache repair, flutter clean and nothing worked. also tried doing a clean VSC install, guess it was some kind of cache problem?

I can't think of a reason that would fix it, but if you see it again, please file a new issue with all the details and a code sample.


> i noticed that the evaluation request registered in VSC did not match the expression i was trying to evaluate at the time

Please also file a new issue with details/instructions to repro and I'll take a look. Thanks!

**Comment by bahadirarslan** [comment]
Still the same at Dart 3.1.1
It pops out somewhere and disappears after a while but I cannot debug when I need.

In my case, for this time I get it from ObjectBox query result. I would like to share a sample repo but it is happening in my current project so I can not reproduce it in a sample app.

 Flutter (Channel stable, 3.13.3, on macOS 13.4.1 22F770820d darwin-arm64, locale en-TR)
    • Flutter version 3.13.3 on channel stable at /Users/bahadirarslan/Development/SDKS/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2524052335 (8 days ago), 2023-09-06 14:32:31 -0700
    • Engine revision b8d35810e9
    • Dart version 3.1.1
    • DevTools version 2.25.0

**Comment by lirantzairi** [comment]
I'm also still getting this issue with the new stable Flutter 3.13.4 and Dart 3.1.2. Reproducible with the example from https://github.com/dart-lang/sdk/issues/52430#issuecomment-1562738203

**Comment by DanTup** [comment]
If this is still happening it needs looking at in the VM - please file an issue at https://github.com/dart-lang/sdk with a repro (or a link to a previously posted repro if it's the same). Please include a screenshot and detailed of exactly where you're seeing this (and what you're hovering over), and the SDK versions you're using.

And please post a link back here. Thanks!

**Comment by rafaellop** [comment]
I'm also unable to do any productive debugging with the OptimizedOut and unavailable values of variables on breakpoints. I can't watch variables or preview them using mouse over or "Evaluate in console". This is super annoying.

```
[√] Flutter (Channel stable, 3.13.5, on Microsoft Windows [Version 10.0.22000.2416])
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
[√] Chrome - develop for the web
[√] Android Studio (version 3.2)
[√] Android Studio (version 2021.3)
[√] VS Code (version 1.82.2)
[√] Connected device (4 available)
[√] Network resources
```

![image](https://github.com/Dart-Code/Dart-Code/assets/8421184/6dad8c72-1a51-439d-91ae-22dfe662d5ba)

![image](https://github.com/Dart-Code/Dart-Code/assets/8421184/0f2d6a48-7ace-491b-b1fd-984dfb704f63)

**Comment by DanTup** [comment]
If you're still seeing this, please add repos to https://github.com/dart-lang/sdk/issues/53654. I'll lock this one because this isn't something that can be solved here and additional comments will just make it less likely for people to see my pointer to the SDK.

This issue can be very dependant on your specific code, so sharing _exact code samples_ that trigger it would likely be very helpful to track it down and fix.
